### ISSUE TITLE
Use Hard Float eabi

### ISSUE THREAD

**Issue by ekiwi** [antoinealb/rust-demo-cortex-m4#1]
Hi Antoin,

it's great to see that you have been toying with Rust on Cortex Microcontrollers as well.
I recently succeeded in linking rust code with the RIOT-OS project runtime. They use hard float EABI calling conventions and it took me some time to figure out how to do this.

My thumbv7em-none-eabi**hf**.json looks like this:

``` .json
{
    "arch": "arm",
    "cpu": "cortex-m4",
    "data-layout": "e-m:e-p:32:32-i1:8:32-i8:8:32-i16:16:32-i64:64-v128:64:128-a:0:32-n32-S64",
    "disable-redzone": true,
    "executables": true,
    "llvm-target": "thumbv7em-none-eabihf",
    "morestack": false,
    "os": "none",
    "relocation-model": "static",
    "target-endian": "little",
    "target-pointer-width": "32",
    "features": "+vfp4,+d16,+fp-only-sp"
}
```

I've successfully tested calling this function from C:

``` .rs
#[no_mangle]
pub fn rust_float_test(x: f32) -> f32 {
    x * x / 2f32
}
```

In case you are wondering how I determined that I had to use `vfp4,+d16,+fp-only-sp` as feature flags: With gcc you can use `-mfpu=fpv4-sp-d16` to specify the FPU of the cortex-m4. Searching for `llvm fpv4-sp-d16` gets you this [https://llvm.org/bugs/attachment.cgi?id=12896](https://llvm.org/bugs/attachment.cgi?id=12896) which led me to believe, that the three features mentioned above need to be enabled.

Thus this involved a lot of guessing and it would be cool if you could try out the target specification and see if it is working with your platform as well.

Kevin

**Comment by antoinealb** [comment]
Hey Kevin,
I will definitely test this, maybe only after Eurobot ;) Just out of curiosity do you know why they use the hardfloat ABI?

Also did you manage to port Rust's `libstd` to your runtime ? That would be really cool to have since a lot of existing code depends on it.

Antoine

**Comment by ekiwi** [comment]
> I will definitely test this, maybe only after Eurobot

Haha, yes. I'm taking a break from Eurobot this year, because I'm working on my bachelors thesis. Good luck to your team!

> Just out of curiosity do you know why they use the hardfloat ABI?

I do not have a definite answer to this. I just guess, that if you compile all your sources at once you do not have to care about binary compatibility. So why not? Is hardfloat ABI inefficient if you are calling functions that do not use floats? For functions using floats performance should be better, right?

> Also did you manage to port Rust's libstd to your runtime ?

No. I've only been using `libcore`, just like you. While I think, that some functionality that requires allocation makes sense on microcontrollers, I have a feeling, that a lot of rust libraries, that exist would not work too well on a deeply embedded target, even if we had `libstd` support.
For my thesis I'm working on a parser, that is supposed to run on my PC as well as on the cortex-m and I find myself having to put a lot of thought into figuring out how to do reach this goal. For example: while you want to have human readable error messages on your computer, on an embedded target error numbers might need to be enough because of memory constraints (I only get 128k of flash for a lot of functionality).

One thing I want to work on is to try and have rust unittests running on the target. However, I'm not sure if I'll be able to make this work, because the test harness stuff seemed to be pretty hard coded into rustc. Also, since it's not the primary objective of my thesis I cannot put too much time into this.

Kevin

**Comment by antoinealb** [comment]
Hello Kevin,
I am sorry for the long delay, I have been kinda busy recently. I finally managed to test your suggestions for using the hardfloat ABI and it worked perfectly (well, once I enabled the FPU...). You can find the needed changes in the `feature-hardfloat-abi` branch.

Did you find a way to run Rust's unit tests on hardware ? Maybe we could open a separate issue for that.

Antoine

**Comment by ekiwi** [comment]
> You can find the needed changes in the feature-hardfloat-abi branch

Great! Thank you. I might have a TIVA launchpad floating around somewhere. I'll check if I get home and try to run your application.

> Did you find a way to run Rust's unit tests on hardware ? Maybe we could open a separate issue for that.

Unfortunately I haven't gotten any further on that issue. I tried to get some brainstorming going with this repository: https://github.com/ekiwi/utest-rs but that did not work out.

The biggest issue right now is how to hook into the rust compiler and the unit test library to provide a different backend.

What do you propose?

**Comment by antoinealb** [comment]
Sadly I don't know enough about Rust's unit testing strategy for the moment, and I never tried to run unit tests on hardware (even when doing C), I always built them on PC :/
